pkg/eventemitter/eventemitter.go

```diff
@@ -0,0 +1,247 @@
+	log.V(1).Info("Start CloudEvent loop.")
+	go e.startEventLoop(ctx)
+	return nil
+}
+
+func (e *EventEmitter) DeleteCloudEvents(cloudEvent *kedav1alpha1.CloudEvent) error {
+	key := cloudEvent.GenerateIdentifier()
+	result, ok := e.eventLoopContexts.Load(key)
+	if ok {
+		cancel, ok := result.(context.CancelFunc)
+		if ok {
+			cancel()
+		}
+		e.eventLoopContexts.Delete(key)
+		e.clearEventHandlersCache(cloudEvent)
+	} else {
+		log.V(1).Info("CloudEvent was not found in controller cache", "key", key)
+	}
+
+	return nil
+}
+
+func (e *EventEmitter) createEventHandlers(ctx context.Context, cloudEvents *kedav1alpha1.CloudEvent) {
+	e.eventHandlersCachesLock.Lock()
+	defer e.eventHandlersCachesLock.Unlock()
+
+	key := cloudEvents.GenerateIdentifier()
+
+	clusterName := cloudEvents.Spec.ClusterName
+	if clusterName == "" {
+		clusterName = "default"
+	}
+
+	if cloudEvents.Spec.Destination.CloudEventHTTP != nil {
+		var err error
+		var eventHandler EventDataHandler
+		eventHandler, err = NewCloudEventHTTPHandler(ctx, clusterName, cloudEvents.Spec.Destination.CloudEventHTTP.Uri, log)
+
+		if err != nil {
+			log.Error(err, "create cloudevent handler failed")
+		}
+		e.eventHandlersCache[key+CloudEventHTTP] = eventHandler
+	}
+}
+
+func (e *EventEmitter) clearEventHandlersCache(cloudEvents *kedav1alpha1.CloudEvent) {
+	e.eventHandlersCachesLock.Lock()
+	defer e.eventHandlersCachesLock.Unlock()
+
+	key := cloudEvents.GenerateIdentifier()
+
+	if cloudEvents.Spec.Destination.CloudEventHTTP != nil {
+		eventHandlerKey := key + CloudEventHTTP
+		if eventHandler, found := e.eventHandlersCache[eventHandlerKey]; found {
+			eventHandler.CloseHandler()
+			delete(e.eventHandlersCache, key)
+		}
+	}
+}
+
+func (e *EventEmitter) startEventLoop(ctx context.Context) {
+	consumingInterval := 500 * time.Millisecond
+
+	if ch == nil {
+		ch = make(chan EventData, MaxChannelBuffer)
+	}
+
+	for {
+		tmr := time.NewTimer(consumingInterval)
+
+		select {
+		case <-tmr.C:
+			tmr.Stop()
+		case eventData := <-ch:
+			log.V(1).Info("Consuming events in queue.")
+			e.emitEventByHandler(eventData)
+		case <-ctx.Done():
+			tmr.Stop()
+			return
+		}
+	}
+}
+
+// Emit is emitting event to both local kubernetes and custom CloudEvents handler. After emit event to local kubernetes, event will inqueue and waitng for handler's consuming.
+func (e *EventEmitter) Emit(object runtime.Object, namesapce types.NamespacedName, eventtype, reason, message string) {
+	e.EventRecorder.Event(object, eventtype, reason, message)
+	name, _ := meta.NewAccessor().Name(object)
+	eventData := EventData{
+		namespace:  namesapce.Namespace,
+		objectName: name,
+		eventtype:  eventtype,
+		reason:     reason,
+		message:    message,
+		time:       time.Now().UTC(),
+	}
+	go e.inqueueEventData(eventData)
+}
+
+func (e *EventEmitter) inqueueEventData(eventData EventData) {
```

Comment:
Sorry, I don't get the logic of this function. We try to enqueue the event, if we can, everything is fine, but if we cannot, we retry it X times until MaxChannelBuffer. How are related the amount of tries and the buffer size?

Reply:
They are separate concepts. MaxChannelBuffer is used to prevent blocking when multiple events need to be sent at the same time, while RetryTimes is used for retrying in case the event sink endpoint is down. I haven't considered the relationship between MaxChannelBuffer and RetryTimes. Do you think there should be some relation between them?